### PR TITLE
Fix buffer size for WS2812 PWM driver

### PR DESCRIPTION
**Pull request by yiancar**
In a recent build, I was trying to use more than 60 ws2812 using the PWM driver (per key RGB).
My STM32F072 run out of RAM. (heap linker error).
Apparently the buffer for the PWM overflow value is a 32bit register...
Did some math to figure out if it need to be this big:
High period for a 1 PWM frequency (CPUCLK/2) / (1000000000 / 800).
Essentially we want to find how many PWM ticks it will take to get 800ns

The above equation never seems to go above 50 ticks... 48MHz CLK is about 19 ticks.

As the DMA controller allows a transfer of a byte of memory into a word of peripheral we better fo that as well:)

Now our buffer is about 1.5Kb of RAM instead of 6.5 Kb.

## Description

Changed DMA for PWM WS2812 driver from word to word TO byte to word.
Changed PWM buffer to a byte instead of a word as it was mostly empty:)

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

*  STM32F072 runs out of RAM with more than 60ish WS2812 LEDs using the PWM driver.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by sigprof** [comment]
> As the DMA controller allows a transfer of a byte of memory into a word of peripheral we better fo that as well:)

The problem is that not all STM32 chips have DMA controllers that support that kind of operation. Older chips (DMAv1 in ChibiOS) can zero-extend data exactly like you want if you specify different MSIZE and PSIZE values. However, some newer chips like F401/F411 (DMAv2 in ChibiOS) cannot do that — depending on DMA FIFO settings, they either force MSIZE and PSIZE to be the same, or perform multiple transfers (either collecting the larger destination data unit by performing multiple reads from the source, or splitting the larger source data unit into multiple writes to the destination). Therefore blindly changing MSIZE to BYTE would break the driver at least on F4x1 chips (and probably the whole F4xx series, or possibly even more).

So the suggested optimization would need to be conditional — the type of DMA controller implementation in the chip needs to be checked somehow. At least the most resource-restricted chips tend to have DMAv1, so the optimization would be possible for them.

For DMAv2 controllers the only possible optimization would be using 16-bit data if the chosen timer is 16-bit (in this case the DMA controller would perform 16-bit writes to the timer registers, which is not formally correct according to the datasheet, but seems to work in practice). Doing the same thing for a 32-bit timer, however, would break things (apparently a 16-bit write to a 32-bit register results in the 16-bit value getting duplicated to both halves of the 32-bit register).

File: platforms/chibios/drivers/ws2812_pwm.c [hunk 1]
```diff
@@ -247,7 +253,7 @@
 
 /* --- PRIVATE VARIABLES ---------------------------------------------------- */
 
-static uint32_t ws2812_frame_buffer[WS2812_BIT_N + 1]; /**< Buffer for a frame */
+uint8_t ws2812_frame_buffer[WS2812_BIT_N + 1]; /**< Buffer for a frame */
```
Comment (drashna) [inline platforms/chibios/drivers/ws2812_pwm.c:256]: Should probably stay static? 
```suggestion
static uint8_t ws2812_frame_buffer[WS2812_BIT_N + 1]; /**< Buffer for a frame */
```

**Comment by yiancar** [comment]
Good shout @sigprof. I can condition this no problem.
How about doing what was done for F1XX SPI like here:
https://github.com/qmk/qmk_firmware/blob/88fe5c16a5cdca5e3cf13ef3cd91f5f1e4898c37/platforms/chibios/chibios_config.h
Define DMAv2 in case of F401/F411. Then if DMAv2 is defined, in the WS2812 driver use 32bit transfers.

Any other suggestions are welcome:)

**Comment by yiancar** [comment]
I did a quick skim of the F401 datasheet. As we are using double buffer mode it seems to be OK. Please look at 9.3.10 of https://www.st.com/resource/en/reference_manual/rm0368-stm32f401xbc-and-stm32f401xde-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

**Comment by sigprof** [comment]
> I did a quick skim of the F401 datasheet. As we are using double buffer mode it seems to be OK. Please look at 9.3.10 of https://www.st.com/resource/en/reference_manual/rm0368-stm32f401xbc-and-stm32f401xde-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

No, see this table — with memory width 8 and peripheral width 32 the controller will do 4 1-byte memory transfers, assemble a 32-bit word from those 4 bytes and perform a single 32-bit peripheral transfer

**Comment by sigprof** [comment]
For comparison, https://www.st.com/resource/en/reference_manual/rm0091-stm32f0x1stm32f0x2stm32f0x8-advanced-armbased-32bit-mcus-stmicroelectronics.pdf has this:



Here the DMA controller zero-extends the data in case of width mismatch, which would be the desired behavior for the WS2812 driver. Unfortunately, the F401 DMA controller does not have an equivalent mode.

**Comment by yiancar** [comment]
Cheers thanks for the clarification:)
I also remembered that this driver supports all timers.. but not all timers are 32 bit:)
So plan of action
1. Use Chibios define `STM32_TIMx_IS_32BITS` to figure out what the peripheral is like. For MCUs with zero-extend do make the necessary transfer (byte to half word/word)
2. Use the defines `STM32F2XX STM32F4XX STM32F7XX` to check if the DMA controller cannot do the zero extend and do so in software, either to a half word or a word.

How does that sound?

**Comment by yiancar** [comment]
@sigprof let me know what you think of it now:)

File: platforms/chibios/drivers/ws2812_pwm.c [hunk 2]
```diff
@@ -29,6 +29,16 @@
...
 
+/* Summarize https://www.st.com/resource/en/application_note/an4013-stm32-crossseries-timer-overview-stmicroelectronics.pdf to
+ * figure out if we are using a 32bit timer. This is needed to setup the DMA controller correctly.
+ * Ignore STM32H7XX and STM32U5XX as they are not supported by ChibiOS.
+ */
+#if (WS2812_PWM_DRIVER == PWMD2) && (!defined(STM32F1XX) || !defined(STM32L0XX) || !defined(STM32L1XX))
+#    define WS2812_PWM_TIMER_32BIT
+#elif (WS2812_PWM_DRIVER == PWMD5) && (!defined(STM32F1XX))
+#    define WS2812_PWM_TIMER_32BIT
+#endif
```
Comment (sigprof) [inline platforms/chibios/drivers/ws2812_pwm.c:40]: The `WS2812_PWM_DRIVER == PWMD2` comparison does not work at the preprocessor level (looks like it is always evaluated as true, so the resulting code still works, but the optimization for 16-bit timers on F4xx never gets enabled). Also the logic in the second part is wrong (it needs `&&`, not `||`).

The following code seems to work for me:

```suggestion
#if !defined(STM32F1XX) && !defined(STM32L0XX) && !defined(STM32L1XX)
#    define WS2812_PWM_TIMER_32BIT_PWMD2 1
#endif
#if !defined(STM32F1XX)
#    define WS2812_PWM_TIMER_32BIT_PWMD5 1
#endif
#define WS2812_CONCAT1(a, b) a##b
#define WS2812_CONCAT(a, b) WS2812_CONCAT1(a, b)
#if WS2812_CONCAT(WS2812_PWM_TIMER_32BIT_, WS2812_PWM_DRIVER)
#    define WS2812_PWM_TIMER_32BIT
#endif
```

File: platforms/chibios/drivers/ws2812_pwm.c [hunk 3]
```diff
@@ -247,13 +263,33 @@
 
 /* --- PRIVATE VARIABLES ---------------------------------------------------- */
 
+// STM32F2XX, STM32F4XX and STM32F7XX do NOT zero pad DMA transfers of unequal data width. Buffer width must match TIMx CCR.
+// For all other STM32 DMA transfer will automatically zero pad. We only need to set the right peripheral width.
+#if defined(STM32F2XX) || defined(STM32F4XX) || defined(STM32F7XX)
+#    if defined(WS2812_PWM_TIMER_32BIT)
+#        define WS2812_DMA_MEMORY_WIDTH STM32_DMA_CR_MSIZE_WORD
+#        define WS2812_DMA_PERIPHERAL_WIDTH STM32_DMA_CR_PSIZE_WORD
 static uint32_t ws2812_frame_buffer[WS2812_BIT_N + 1]; /**< Buffer for a frame */
```
Comment (sigprof) [inline platforms/chibios/drivers/ws2812_pwm.c:272]: Maybe add a typedef or something, and then define `ws2812_frame_buffer` only once?

**Review by sigprof — CHANGES_REQUESTED** [review]
The code with my changes seems to work on STM32F401; tested pin B15 with TIM1 (with 16-bit optimization enabled) and pin A2 with TIM5 (in this case using a 16-bit data buffer fails, 32-bit works as before). Still need to test on some other chips.

File: platforms/chibios/drivers/ws2812_pwm.c (hunk 2 above)
Comment (yiancar) [reply platforms/chibios/drivers/ws2812_pwm.c:40]: After a good conversation I learned that comparison in defines in the C preprocessor will only work correctly for numerical values. Proposed code seems to be the cleanest.

**Comment by yiancar** [comment]
Hello!
Any other changes you would like?

**Review by tzarc — APPROVED** [review]

**Comment by drashna** [comment]
Sorry, has a merge conflict

**Review by drashna — APPROVED** [review]
Can confirm it doesn't break anything

**Pull request merged**